bpf/lib/nat.h

```diff
@@ -711,11 +711,15 @@ static __always_inline void snat_v4_init_tuple(const struct iphdr *ip4,
  * - target->egress_gateway to true if the packet should be SNAT-ed because of
  *   an egress gateway policy
  *
- * The function will return true if the packet should be SNAT-ed, false
+ * The function will also set *snat_needed to true if the packet should be
+ * SNAT-ed, to false otherwise.
+ *
```

Comment:
Note that one common pattern in the NAT code is to return `NAT_PUNT_TO_STACK` for "no error, but also no NAT handling needed".

Not sure if this would make things much easier here.

Reply:
Do you mean, instead of returning an `int` plus passing the bool pointer? I hesitated between the two solutions, having error on < 0, no SNAT on 0 (or on `NAT_PUNT_TO_STACK`), and SNAT on > 0 (or >= 0). Passing the bool pointer is more verbose in `snat_v*_prepare_state()`, but I thought it might be easier to follow. I don't mind switching, though, if you think it would be more consistent?

Reply:
> Do you mean, instead of returning an `int` plus passing the bool pointer?

Yep, exactly.

> Passing the bool pointer is more verbose in `snat_v*_prepare_state()`, but I thought it might be easier to follow.

Imho part of the current verbosity comes from all the (redundant?) `*snat_needed = false;` statements. If we expect the caller to initialize with `false`, do we still need these?

> I don't mind switching, though, if you think it would be more consistent?

Given that the callers already have awareness of what `NAT_PUNT_TO_STACK` means, it does feel a bit more consistent. But I know this code in way too much detail by now, hard to say what's easier to grasp on a casual glance :).

Reply:
OK let's use `NAT_PUNT_TO_STACK` then. I'll use the comment on top of the IPv4 function to describe what the expectations are about the return value.